### ISSUE TITLE
Attempt 03: heavier decoder with learned upsampling on cached Virchow2 features

### ISSUE THREAD

**Issue by clemsgrs** [clemsgrs/beetle-solution#5]
## Agreed experiment

Use Soma's existing `heavy_conv` decoder with width 256, two learned transposed-convolution upsampling blocks, GroupNorm with 32 groups, and pyramid-pooling scales [1, 2, 3, 6]. The user approved this design, the Attempt 03 identifier, and implementation/preflight preparation on 2026-09-08.

The decoder has 5,251,588 parameters versus Attempt 01's 1,510,660 and Attempt 02's 2,691,844. Two blocks preserve Attempt 01's native 148×148 logit resolution; the established bilinear resize to 518×518 and top-left crop to 512×512 remain. This tests context pooling, learned upsampling, and greater capacity together, rather than isolating the upsampling operator. It adds no RGB or intermediate-layer encoder features.

The exact pinned Soma revision `4a3d6c84f9a3dc2832e585c6884a95d14a0f79bc` already implements this decoder. No dependency upgrade or new decoder implementation is needed.

## Cache and protocol

Reuse the exact frozen Virchow2 cache from Attempts 01/02:

```text
/maindisk/clement/soma-beetle-campaign-20260827/data/beetle/cache/virchow2_3158645804b69e3f3bc4439d4116edddf0840a72_8d6cea947eb2418c3b0dff48cfb9b238e47744ab0dfca21b2b0637b140769b4b_dense_fp16/dense/95ab55f548038c00/dense_embeddings
```

The source identity is locked by `configs/attempts/attempt-02-cache-lock.json`. The full audit verified all 124,697 FP16 tensors and 124,697 sidecars, 437,290,638,590 tensor bytes, and the dataset/split hashes against the immutable manifest. No features were extracted or replaced.

Preserve the existing five-fold development protocol, uniform ROI sampling, seed, loss, optimizer, learning rate, schedule, patience, checkpoint selection, effective batch size 64, and per-fold ROI draw budgets. Attempt 01 is the primary comparator because decoder depth and native output resolution match; Attempt 02 is secondary.

**Evaluation interpretation:** curation assigns test=k, tune=(k+1)%5, train=the other three folds. With `holdout_test: true`, the declared test fold is excluded. These are development/tune scores used for checkpoint selection, not independent test estimates. A protocol change requires a matched baseline and belongs in a separate experiment.

## Completed preparation

- [x] Agreed architecture, scope, comparators, and Attempt 03 identity.
- [x] Added `configs/attempts/attempt-03.yaml` and `beetle/attempt_03.py` locally; the protocol check permits only the agreed decoder change apart from run metadata.
- [x] Rehashed every tensor and sidecar and verified exact cache coverage and dataset/split identity.
- [x] Exercised a real cached feature tensor with synthetic four-class masks and ignored pixels through the pinned Soma decoder/head in FP32. Checked finite loss/gradients, Adam updates, 512×512 output geometry, and exact checkpoint reload.
- [x] All four GPU candidates passed; selected **64×1**, frozen across all five folds.
- [x] Generated the absolute-path launch configuration: `provenance/attempts/attempt-03/launch/attempt-03.yaml`.
- [x] Full repository test suite: **49 passed in 24.44 seconds**, using an archive of exact repository commit `5ca2fe865cde41fd45008c7622126627caa2b23b` plus the verified Attempt 03 changes. Local source copies avoided stalled repository filesystem lookups.

| Physical batch × accumulation | Peak allocated GiB | Peak reserved GiB |
|---|---:|---:|
| 64×1 | 10.99 | 17.29 |
| 32×2 | 5.56 | 7.37 |
| 16×4 | 2.82 | 4.41 |
| 8×8 | 1.47 | 1.93 |

Measurements used an NVIDIA H200. Throughput is from a short synthetic-target compute probe, not end-to-end training. Runtime/source provenance and preflight reports are under `provenance/attempts/attempt-03/`; the experiment record and commands are in `docs/attempt-03.md`. The implementation and evidence currently exist in the working tree; this issue does not imply they have been committed or pushed.

## Remaining experiment work

**Five-fold run launched 2026-09-08; initializing.** Run `attempt-03-heavy-conv-20260908-r1` is detached under `/maindisk/clement/beetle-attempt-03-20260908`. Optimizer steps have not yet been confirmed. See launch update below.

- [ ] Run all five folds independently of interim scores, retaining checkpoints, resolved configurations, histories, sampling audits, confusion evidence, environment provenance, and checksums.
- [ ] Report equal-weight mean and sample standard deviation of five fold dataset-global four-class macro Dice scores, paired deltas against Attempt 01, and Attempt 02 as a secondary reference.
- [ ] Retain per-class and pixel-micro Dice and paired 527-patient primary / 524-patient sensitivity summaries; inspect representative boundaries and periodic upsampling artifacts.
- [ ] Publish the evidence and checkpoints and record the conclusion regardless of score. A null or negative result completes the experiment; External model selection is separate.

Related: #2 (completed four-block lightweight-decoder ablation).

**Comment by clemsgrs** [comment]
The user authorized the full five-fold run on 2026-09-08. The detached process is currently initializing; optimizer steps are not yet confirmed.

- Run: `attempt-03-heavy-conv-20260908-r1`; PID 4326 at launch.
- Campaign: `/maindisk/clement/beetle-attempt-03-20260908`; outputs: `runs/`; live state: `status.json`; append-only log: `training.log`.
- Exact Soma 1.11.2 pinned source installed in an isolated environment and source checksum verified at startup. Heavy decoder, batch 64 × accumulation 1, all five folds and approved scientific settings retained.
- Original campaign output directory is unwritable by the current uid. Original ROI coordinates and population metadata were mode 0600 under uid 10170. Initial startup attempts failed before training; their logs remain available.
- Writable metadata cache wrapper points its dense directory directly to the original Attempt 01/02 feature cache. ROI coordinates recovered from Attempt 02's archived manifest reproduce byte-identical ROI manifest/split CSVs and all 124,697 payload identities. No ROI resampling or feature extraction was performed by the active run.
- Startup confirms **587 slides from cache, zero slides to sample**, and resolves original dense cache key **95ab55f548038c00**, covering 124,697 artifacts. Payload validation completed successfully: Soma reports a complete feature-cache hit and reuse of cached ROI dense grids. The process is now rebuilding per-ROI class counts from the original masks; optimizer steps are not yet confirmed. The user has been given a read-permission command so the remaining population metadata can also be reused.

Launch, runtime, cache-access recovery, active configuration, and status snapshots are saved under `provenance/attempts/attempt-03/` in the working tree. Live campaign state takes precedence over the snapshot. No training metrics are available yet.

**Comment by clemsgrs** [comment]
Training confirmed on 2026-09-08: the active `attempt-03-heavy-conv-20260908-r1` process has progressed into fold 0 and saved `fold_0/best_model.pt`. The checkpoint loads successfully and records epoch 2 with tune loss 0.7336154609746823. This is the best saved checkpoint epoch, not a measurement of the current live epoch. The detached process remains active. All 124,697 ROI class-count records were rebuilt and published successfully before training; the original dense feature cache was reused. The detached text log does not emit live epoch progress.

**Comment by clemsgrs** [comment]
Attempt 03 was temporarily stopped at the user’s request on 2026-09-08 because the GPU is shared with another active workload. Sent SIGTERM only to our verified, isolated training process group (4326). Confirmed no remaining group members and that our GPU allocation disappeared; the other GPU process remains running. Live status is `paused`, with automatic restart disabled. Existing checkpoints and completed fold artifacts are preserved. No relaunch will occur until the user requests it.

**Comment by clemsgrs** [comment]
**Result.** All five folds of `attempt-03-heavy-conv-20260908-r1` completed on 2026-09-10. Dev Dice 0.8894 ± 0.0286 (sample SD) versus Attempt 01 0.8861 ± 0.0285: paired +0.0033 with 4/5 folds improved, patient-bootstrap Δ +0.0046 [−0.0022, +0.0116] on 527 patients (524-patient sensitivity cohort: +0.0046 [−0.0025, +0.0114]). Versus Attempt 02: +0.0014, 3/5 folds. Pooled class Dice versus Attempt 01: non-invasive +0.0066, invasive +0.0045, necrosis +0.0079, other −0.0008.

**Verdict.** A small signal against Attempt 01 and no detectable effect against Attempt 02. Decoder capacity on the frozen token grid is a low-yield axis. Development-only; the Attempt 01 External model stands.

**Qualitative check.** Transposed upsampling leaves a sub-visible 7-px boundary phase bias (max/min 1.28 versus 1.03 for Attempt 02's bilinear decoder); there is no visible checkerboard.

**Deviations.**
- ROI coordinate and population files were unreadable (mode 0600, another uid). Coordinates were rebuilt from Attempt 02's archived ROI manifest (byte-identical manifest and splits) and class counts from the masks; features were reused, not re-extracted.
- The shared GPU forced three stops: a manual stop on 09-08 during fold 1, and guard yields on 09-10 at 09:05 and 15:09 UTC during fold 4. Soma has no epoch-level resume, so folds 1 and 4 were rerun from epoch 0. Fold 1's interrupted run had reached 0.9106; the reported 0.9086 is from the complete rerun.

Evidence: #6, release [`attempt-03`](https://github.com/clemsgrs/beetle-solution/releases/tag/attempt-03), `docs/research-journal.md`.
